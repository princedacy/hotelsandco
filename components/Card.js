import React from "react";
import Image from "next/image";
import { StarIcon } from "@heroicons/react/solid";
import { useQuery } from "@tanstack/react-query";
function Card({router}) {
  const { data, isLoading, error } = useQuery({
    queryKey: ["rooms"],
    queryFn: () =>
      fetch("https://63afdf4f649c73f572c415b1.mockapi.io/api/rooms").then(
        (res) => res.json()
      ),
  });
  if (isLoading) return "Loading...";

  if (error) return "An error has occurred: " + error.message;

  const openRoom = async (room) => {
    await router.push({
      pathname: "/room",
      query: {
        id: room.id,
        location: room.location,
        image: room.image,
        price: room.price,
        rating: room.rating,
        description: room.description,
        reviews: room.reviews,
        guests: room.guests,
        bedrooms: room.bedrooms,
        baths: room.baths,
        beds: room.beds,
        host_image: room.host_image,
        host_name: room.host_name,
        city: room.city
      },
    });
  };
  return (
    <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-4 gap-1 px-20">
      {data.map((room) => (
        <div
          className="card-container container mx-auto relative w-80 h-80 mb-16 cursor-pointer"
          key={room.id}
        >
          <div
            className="card group text-left p-3 mb-5 relative w-full"
            onClick={() => openRoom(room)}
          >
            <Image
              className="!w-5 !h-5 absolute !top-5 !right-5 cursor-pointer z-40 !left-auto"
              src={"/icons/heart.svg"}
              alt="add to wishlist"
              style={{ objectFit: "contain", objectPosition: "right" }}
              fill
            />
            <Image
              src={room.image}
              alt=""
              style={{ objectFit: "contain", objectPosition: "center" }}
              fill
              className="!relative rounded-2xl"
            />
            <div className="flex items-center justify-between">
              <h3 className="location-name text-black font-medium py-2">
                {room.location}
              </h3>
              <div className="flex items-center">
                {(room.rating * 0.5) >= 3 ? (
                  <>
                    <StarIcon className="h-4" />
                    <p className="font-light text-sm">{room.rating * 0.5}</p>
                  </>
                ) : (
                  ""
                )}
              </div>
            </div>
            <h3 className="text-gray-500 font-light">{room.city}</h3>
            <p className="py-2 text-sm text-gray-500 font-light">Jan 1 â€“ 6</p>
            <div className="flex">
              <span className="font-regular">${room.price}&nbsp;</span>
              <span className="font-light">night</span>
            </div>
          </div>
        </div>
      ))}
    </div>
  );
}

export default Card;
