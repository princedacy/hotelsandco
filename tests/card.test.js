import React from "react";
import { render, screen, act, waitFor } from "@testing-library/react";
import ServiceWorker from "../mocks/service-worker";
import Card from "../components/Card";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

global.window = {};

const queryClient = new QueryClient();

describe("Card", () => {
  beforeEach(() => {
    ServiceWorker.mockClear();
    if (!window.navigator.serviceWorker) {
      window.navigator.serviceWorker = {};
    }
    window.navigator.serviceWorker.register = jest.fn();
  });

  it("should render the component and take a snapshot", async () => {
    jest.setTimeout(20000);
    const originalUseQuery = React.useQuery;
    React.useQuery = jest.fn(() => ({
      data: [
        {
          number: 89025,
          status: false,
          wifi: true,
          location: "Rogers",
          image: "https://loremflickr.com/640/480/city",
          price: "517.00",
          rating: "1",
          description:
            "Quisquam in cumque. Laboriosam ipsum dignissimos consequuntur iusto ipsum blanditiis. Iusto expedita provident pariatur minima tempora quidem assumenda corporis. Sequi illum voluptatum impedit voluptatum voluptatibus dolores quibusdam nulla quis. Repellendus est voluptates officia commodi nihil vel quaerat adipisci doloremque. Tempore laboriosam earum eum veniam reiciendis non culpa.",
          reviews: "2",
          host_image:
            "https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/72.jpg",
          guests: "1",
          bedrooms: "1",
          beds: "2",
          baths: "3",
          host_name: "Mr. Marc Simonis",
          city: "Southaven",
          id: "1",
        },
      ],
      isLoading: false,
      error: null,
    }));

    const tree = (
      <QueryClientProvider client={queryClient}>
        <Card />
      </QueryClientProvider>
    );
    const { container } = render(tree);
    expect(container).toMatchSnapshot();

    await act(async () => {
      expect(container).toMatchSnapshot();
      await waitFor(() =>
        expect(screen.getByText("Rogers")).toBeInTheDocument()
      );
      await waitFor(() => expect(screen.getByAltText("")).toBeInTheDocument());
      await waitFor(() =>
        expect(screen.getByText("Southaven")).toBeInTheDocument()
      );
      await waitFor(() =>
        expect(screen.getByText("Jan 1 â€“ 6")).toBeInTheDocument()
      );
      await waitFor(() =>
        expect(screen.getByText("$517 night")).toBeInTheDocument()
      );
    });

    React.useQuery = originalUseQuery;
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
