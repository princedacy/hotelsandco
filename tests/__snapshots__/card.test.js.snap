// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render the component and take a snapshot 1`] = `
<div>
  Loading...
</div>
`;

exports[`Card should render the component and take a snapshot 2`] = `
<div>
  Loading...
</div>
`;
